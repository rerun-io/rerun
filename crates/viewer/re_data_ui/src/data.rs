use re_viewer_context::{UiLayout, ViewerContext};

use super::DataUi;

impl DataUi for re_types::datatypes::Mat3x3 {
    fn data_ui(
        &self,
        _ctx: &ViewerContext<'_>,
        ui: &mut egui::Ui,
        _ui_layout: UiLayout,
        _query: &re_chunk_store::LatestAtQuery,
        _db: &re_entity_db::EntityDb,
    ) {
        egui::Grid::new("mat3").num_columns(3).show(ui, |ui| {
            ui.monospace(self[0].to_string());
            ui.monospace(self[3].to_string());
            ui.monospace(self[6].to_string());
            ui.end_row();

            ui.monospace(self[1].to_string());
            ui.monospace(self[4].to_string());
            ui.monospace(self[7].to_string());
            ui.end_row();

            ui.monospace(self[2].to_string());
            ui.monospace(self[5].to_string());
            ui.monospace(self[8].to_string());
            ui.end_row();
        });
    }
}
