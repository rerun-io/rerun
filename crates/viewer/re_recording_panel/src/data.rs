//! Data structures describing the contents of the recording panel.

use std::collections::BTreeMap;
use std::sync::Arc;
use std::task::Poll;

use ahash::HashMap;
use itertools::Itertools as _;

use re_entity_db::EntityDb;
use re_entity_db::entity_db::EntityDbClass;
use re_log_types::{ApplicationId, EntryId, TableId, natural_ordering};
use re_redap_browser::{Entries, EntryInner, RedapServers};
use re_smart_channel::SmartChannelSource;
use re_types::archetypes::RecordingInfo;
use re_types::components::{Name, Timestamp};
use re_viewer_context::{DisplayMode, Item, ViewerContext};

/// Short-lived structure containing all the data that will be displayed in the recording panel.
#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct RecordingPanelData<'a> {
    /// All the locally loaded application IDs and the corresponding recordings.
    pub local_apps: Vec<AppIdData<'a>>,

    /// All the locally loaded tables.
    pub local_tables: Vec<TableId>,

    /// All the loaded examples
    pub example_apps: Vec<AppIdData<'a>>,

    /// Should the example section be displayed at all?
    pub show_example_section: bool,

    /// All the configured servers.
    pub servers: Vec<ServerData<'a>>,

    /// Recordings that are currently being loaded that we cannot attribute yet to a specific
    /// section.
    pub loading_receivers: Vec<Arc<SmartChannelSource>>,
}

impl<'a> RecordingPanelData<'a> {
    pub fn new(ctx: &'a ViewerContext<'a>, servers: &'a RedapServers, hide_examples: bool) -> Self {
        re_tracing::profile_function!();

        //
        // Find relevant loading sources
        //

        let mut loading_receivers = vec![];
        let mut loading_partitions: HashMap<
            re_uri::Origin,
            HashMap<EntryId, Vec<Arc<SmartChannelSource>>>,
        > = HashMap::default();

        let sources_with_stores: ahash::HashSet<SmartChannelSource> = ctx
            .storage_context
            .bundle
            .recordings()
            .filter_map(|store| store.data_source.clone())
            .collect();

        for source in ctx.connected_receivers.sources() {
            if sources_with_stores.contains(&source) {
                continue;
            }

            match source.as_ref() {
                SmartChannelSource::File(_) | SmartChannelSource::RrdHttpStream { .. } => {
                    loading_receivers.push(source);
                }

                SmartChannelSource::RedapGrpcStream { uri, .. } => {
                    loading_partitions
                        .entry(uri.origin.clone())
                        .or_default()
                        .entry(EntryId::from(uri.dataset_id))
                        .or_default()
                        .push(source);
                }

                // We only show things we know are very-soon-to-be recordings, which these are not.
                SmartChannelSource::RrdWebEventListener
                | SmartChannelSource::JsChannel { .. }
                | SmartChannelSource::Sdk
                | SmartChannelSource::Stdin
                | SmartChannelSource::MessageProxy(_) => {}
            }
        }

        //
        // Find everything else
        //

        let servers = servers
            .iter_servers()
            .map(|server| ServerData::new(ctx, server, loading_partitions.get(server.origin())))
            .collect();

        let mut local_apps: BTreeMap<ApplicationId, Vec<&EntityDb>> = Default::default();
        let mut examples_apps: BTreeMap<ApplicationId, Vec<&EntityDb>> = Default::default();

        for entity_db in ctx.storage_context.bundle.entity_dbs() {
            let app_id = entity_db.application_id();
            match entity_db.store_class() {
                EntityDbClass::LocalRecording => local_apps
                    .entry(app_id.clone())
                    .or_default()
                    .push(entity_db),

                EntityDbClass::ExampleRecording => examples_apps
                    .entry(app_id.clone())
                    .or_default()
                    .push(entity_db),

                // these are either handled elsewhere or ignored
                EntityDbClass::DatasetPartition(_) | EntityDbClass::Blueprint => {}
            }
        }

        let local_apps = local_apps
            .into_iter()
            .map(|(app_id_or_examples, entity_dbs)| {
                AppIdData::new(ctx, app_id_or_examples, entity_dbs)
            })
            .collect();

        let example_apps: Vec<_> = examples_apps
            .into_iter()
            .map(|(app_id_or_examples, entity_dbs)| {
                AppIdData::new(ctx, app_id_or_examples, entity_dbs)
            })
            .collect();

        let show_example_section = ctx
            .app_options()
            .include_rerun_examples_button_in_recordings_panel
            && !hide_examples
            || !example_apps.is_empty();

        let local_tables = ctx
            .storage_context
            .tables
            .keys()
            .sorted()
            .cloned()
            .collect();

        Self {
            local_apps,
            local_tables,
            example_apps,
            show_example_section,
            servers,
            loading_receivers,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.local_apps.is_empty()
            && self.local_tables.is_empty()
            && self.example_apps.is_empty()
            && self.servers.is_empty()
    }
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct AppIdData<'a> {
    pub app_id: ApplicationId,
    pub is_active: bool,
    pub is_selected: bool,

    pub loaded_recordings: Vec<RecordingData<'a>>,
}

impl<'a> AppIdData<'a> {
    fn new(
        ctx: &'a ViewerContext<'a>,
        app_id: ApplicationId,
        mut entity_dbs: Vec<&'a EntityDb>,
    ) -> Self {
        entity_dbs.sort_by_cached_key(|entity_db| {
            (
                entity_db
                    .recording_info_property::<Name>(&RecordingInfo::descriptor_name())
                    .map(|s| natural_ordering::OrderedString(s.to_string())),
                entity_db
                    .recording_info_property::<Timestamp>(&RecordingInfo::descriptor_start_time()),
            )
        });

        let is_active = false;
        let is_selected = ctx.selection().contains_item(&Item::AppId(app_id.clone()));

        let loaded_recordings = entity_dbs
            .into_iter()
            .map(|entity_db| RecordingData { entity_db })
            .collect();

        Self {
            app_id,
            is_active,
            is_selected,
            loaded_recordings,
        }
    }

    pub fn id(&self) -> egui::Id {
        egui::Id::new(&self.app_id)
    }

    pub fn name(&self) -> &str {
        self.app_id.as_str()
    }

    pub fn item(&self) -> Item {
        Item::AppId(self.app_id.clone())
    }
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct RecordingData<'a> {
    #[cfg_attr(feature = "testing", serde(serialize_with = "serialize_entity_db"))]
    pub entity_db: &'a EntityDb,
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct ServerData<'a> {
    pub origin: re_uri::Origin,
    pub is_active: bool,
    pub is_selected: bool,

    pub entries_data: ServerEntriesData<'a>,
}

impl<'a> ServerData<'a> {
    fn new(
        ctx: &'a ViewerContext<'_>,
        server: &re_redap_browser::Server,
        loading_partitions: Option<&HashMap<EntryId, Vec<Arc<SmartChannelSource>>>>,
    ) -> Self {
        let origin = server.origin();
        let item = Item::RedapServer(origin.clone());

        let is_selected = ctx.selection().contains_item(&item);
        let is_active = matches!(
            ctx.display_mode(),
            DisplayMode::RedapServer(current_origin)
            if current_origin == origin
        );

        let entries_data =
            ServerEntriesData::new(ctx, server.entries(), origin, loading_partitions);

        Self {
            origin: origin.clone(),
            is_active,
            is_selected,
            entries_data,
        }
    }

    pub fn item(&self) -> Item {
        Item::RedapServer(self.origin.clone())
    }
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub enum ServerEntriesData<'a> {
    Loading,

    Error(String),

    Loaded {
        dataset_entries: Vec<DatasetData<'a>>,
        table_entries: Vec<RemoteTableData>,
        failed_entries: Vec<FailedEntryData>,
    },
}

impl<'a> ServerEntriesData<'a> {
    fn new(
        ctx: &'a ViewerContext<'a>,
        entries: &Entries,
        origin: &re_uri::Origin,
        loading_partitions: Option<&HashMap<EntryId, Vec<Arc<SmartChannelSource>>>>,
    ) -> Self {
        match entries.state() {
            Poll::Ready(Ok(entries)) => {
                let mut dataset_entries = vec![];
                let mut table_entries = vec![];
                let mut failed_entries = vec![];

                for entry in entries.values().sorted_by_key(|entry| entry.name()) {
                    let entry_data = EntryData {
                        origin: origin.clone(),
                        entry_id: entry.id(),
                        name: entry.name().to_owned(),
                        icon: entry.icon(),
                        is_selected: ctx.selection().contains_item(&Item::RedapEntry(
                            re_uri::EntryUri {
                                origin: origin.clone(),
                                entry_id: entry.id(),
                            },
                        )),
                        is_active: ctx.active_redap_entry() == Some(entry.id()),
                    };

                    match entry.inner() {
                        Ok(EntryInner::Dataset(_dataset)) => {
                            let mut displayed_partitions: Vec<PartitionData<'_>> = ctx
                                .storage_context
                                .bundle
                                .entity_dbs()
                                .filter_map(|entity_db| {
                                    if let EntityDbClass::DatasetPartition(uri) =
                                        entity_db.store_class()
                                    {
                                        if &uri.origin == origin
                                            && EntryId::from(uri.dataset_id) == entry.id()
                                        {
                                            Some(PartitionData::Loaded { entity_db })
                                        } else {
                                            None
                                        }
                                    } else {
                                        None
                                    }
                                })
                                .collect();

                            if let Some(loading_partitions) = loading_partitions
                                && let Some(smart_channels) = loading_partitions.get(&entry.id())
                            {
                                displayed_partitions.extend(smart_channels.iter().map(|source| {
                                    PartitionData::Loading {
                                        receiver: source.clone(),
                                    }
                                }));
                            }

                            dataset_entries.push(DatasetData {
                                entry_data,
                                displayed_partitions,
                            });
                        }

                        Ok(EntryInner::Table(_table)) => {
                            table_entries.push(RemoteTableData { entry_data });
                        }

                        Err(err) => failed_entries.push(FailedEntryData {
                            entry_data,
                            error: err.to_string(),
                        }),
                    }
                }

                Self::Loaded {
                    dataset_entries,
                    table_entries,
                    failed_entries,
                }
            }

            Poll::Ready(Err(err)) => Self::Error(err.to_string()),

            Poll::Pending => Self::Loading,
        }
    }
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct DatasetData<'a> {
    pub entry_data: EntryData,
    pub displayed_partitions: Vec<PartitionData<'a>>,
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct RemoteTableData {
    pub entry_data: EntryData,
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct FailedEntryData {
    pub entry_data: EntryData,
    pub error: String,
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub struct EntryData {
    pub origin: re_uri::Origin,
    pub entry_id: re_log_types::EntryId,

    pub name: String,

    #[cfg_attr(feature = "testing", serde(serialize_with = "serialize_icon"))]
    pub icon: re_ui::icons::Icon,

    pub is_selected: bool,
    pub is_active: bool,
}

impl EntryData {
    pub fn item(&self) -> Item {
        Item::RedapEntry(self.entry_uri())
    }

    pub fn id(&self) -> egui::Id {
        egui::Id::new(&self.origin)
            .with(self.entry_id)
            .with(&self.name)
    }

    pub fn entry_uri(&self) -> re_uri::EntryUri {
        re_uri::EntryUri {
            origin: self.origin.clone(),
            entry_id: self.entry_id,
        }
    }
}

// ---

#[derive(Debug)]
#[cfg_attr(feature = "testing", derive(serde::Serialize))]
pub enum PartitionData<'a> {
    Loading {
        receiver: Arc<SmartChannelSource>,
    },
    Loaded {
        #[cfg_attr(feature = "testing", serde(serialize_with = "serialize_entity_db"))]
        entity_db: &'a EntityDb,
    },
}

impl PartitionData<'_> {
    pub fn entity_db(&self) -> Option<&EntityDb> {
        match self {
            PartitionData::Loaded { entity_db, .. } => Some(entity_db),
            PartitionData::Loading { .. } => None,
        }
    }
}

// ---

#[cfg(feature = "testing")]
fn serialize_entity_db<S>(value: &EntityDb, serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    use serde::Serialize as _;
    value.store_id().serialize(serializer)
}

#[cfg(feature = "testing")]
fn serialize_icon<S>(value: &re_ui::Icon, serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    use serde::Serialize as _;
    value.uri().serialize(serializer)
}
