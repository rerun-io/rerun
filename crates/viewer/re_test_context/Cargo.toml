[package]
name = "re_test_context"
description = "A common context used for tests."
authors.workspace = true
edition.workspace = true
homepage.workspace = true
include.workspace = true
license.workspace = true
publish = true
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[lints]
workspace = true

[dependencies]
re_build_info.workspace = true
re_chunk.workspace = true
re_chunk_store.workspace = true
re_entity_db.workspace = true
re_global_context.workspace = true
re_grpc_client.workspace = true
re_log.workspace = true
re_log_encoding.workspace = true
re_log_types.workspace = true
re_renderer.workspace = true
re_tracing.workspace = true
re_types.workspace = true
re_types_core.workspace = true
re_ui.workspace = true
re_viewer_context.workspace = true

# External dependencies
ahash.workspace = true
anyhow.workspace = true
egui-wgpu.workspace = true
egui.workspace = true
egui_kittest.workspace = true
jiff.workspace = true
once_cell.workspace = true
parking_lot.workspace = true
pollster.workspace = true
wgpu.workspace = true
