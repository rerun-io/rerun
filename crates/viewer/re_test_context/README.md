# re_test_context

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_test_context)](https://crates.io/crates/re_test_context?speculative-link)
[![Documentation](https://docs.rs/re_test_context/badge.svg)](https://docs.rs/re_test_context?speculative-link)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

> [!WARNING]
> Used for testing only.

Provides a common context for test cases.
