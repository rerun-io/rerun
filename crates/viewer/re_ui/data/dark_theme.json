{
  "Alias": {
    "Typography": {
      "Default": {
        "value": "{Global.Typography.200}",
        "description": "Default font in Rerun's UI",
        "type": "typography"
      }
    }
  },
  "Global": {
    "Typography": {
      "100": {
        "value": {
          "fontSize": "10px",
          "fontWeight": "Medium",
          "fontFamily": "Inter",
          "lineHeight": "12px",
          "letterSpacing": "-0.12px"
        },
        "description": "",
        "type": "typography"
      },
      "200": {
        "value": {
          "fontSize": "12px",
          "fontWeight": "Medium",
          "fontFamily": "Inter",
          "lineHeight": "16px",
          "letterSpacing": "-0.15px"
        },
        "description": "",
        "type": "typography"
      }
    }
  }
}
