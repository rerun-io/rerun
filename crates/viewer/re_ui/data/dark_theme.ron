{
  "large_button_size": 22,
  "large_button_icon_size": 12,
  "large_button_corner_radius": 6,
  "small_icon_size": 14,

  "Alias": {
    "native_frame_stroke": {
      "color": "{Gray.250}",
      "width": 1
    },
    "top_bar_color": {
      "color": "{Gray.100}"
    },
    "tab_bar_color": {
      "color": "{Gray.200}"
    },
    "bottom_bar_color": {
      "color": "{Gray.150}"
    },
    "bottom_bar_stroke": {
      "color": "{Gray.250}",
      "width": 1
    },
    "shadow_gradient_dark_start": {
      "color": "{Gray.0}",
      "alpha": 77
    },
    "strong_fg_color": {
      "color": "{Gray.1000}"
    },
    "info_log_text_color": {
      "color": "{Green.700}"
    },
    "debug_log_text_color": {
      "color": "{Blue.700}"
    },
    "trace_log_text_color": {
      "color": "{Gray.700}"
    },

    "success_text_color": {
      "color": "{Green.700}"
    },
    "info_text_color": {
      "color": "{Blue.700}"
    },

    "highlight_color": {
      // Background color for widgets that should catch the user's attention
      "color": "{Blue.500}",
      "alpha": 120
    },

    "label_button_icon_color": {
      "color": "{Gray.775}"
    },
    "section_collapsing_header_color": {
      "color": "{Gray.200}"
    },
    "loop_selection_color": {
      "color": "#012569B2"
    },
    "loop_everything_color": {
      "color": "#02502D"
    },
    "thumbnail_background_color": {
      "color": "{Gray.250}"
    },

    "example_card_background_color": {
      "color": "#FFFFFF0A"
    },
    "example_tag_bg_fill": {
      "color": "{Gray.300}",
    },
    "example_tag_stroke": {
      "color": "{Gray.200}",
      "width": 1
    },

    "breadcrumb_text_color": {
      "color": "#6A8CD0"
    },
    "breadcrumb_separator_color": {
      "color": "{Blue.500}"
    },

    "panel_bg_color": {
      "color": "{Gray.100}"
    },

    "blueprint_time_panel_bg_fill": {
      "color": "#141326"
    },

    "notification_panel_background_color": {
      "color": "{Gray.150}"
    },
    "notification_background_color": {
      "color": "{Gray.200}"
    },

    "table_header_bg_fill": {
      "color": "{Gray.150}"
    },
    "table_header_stroke_color": {
      "color": "{Gray.300}"
    },
    "table_interaction_hovered_bg_stroke": {
      "color": "{Gray.300}"
    },
    "table_interaction_active_bg_stroke": {
      "color": "{Gray.350}"
    },
    "table_interaction_noninteractive_bg_stroke": {
      "color": "{Gray.200}"
    },
    "table_sort_icon_color": {
      "color": "{Blue.450}"
    },

    "drag_pill_droppable_fill": {
      "color": "{Blue.325}"
    },
    "drag_pill_droppable_stroke": {
      "color": "{Blue.375}"
    },
    "drag_pill_nondroppable_fill": {
      "color": "{Gray.325}"
    },
    "drag_pill_nondroppable_stroke": {
      "color": "{Gray.375}"
    },
    "drop_target_container_stroke": {
      "color": "{Blue.350}",
      "width": 2
    },
    "tile_drag_preview_stroke" : {
      "color": "{Gray.1000}",
      "alpha": 127,
      "width": 1
    },
    "tile_drag_preview_color": {
      "color": "{Gray.1000}",
      "alpha": 26
    },

    "floating_color": {
      "color": "{Gray.250}"
    },
    "faint_bg_color": {
      "color": "{Gray.150}"
    },
    "extreme_bg_color": {
      "color": "{Gray.0}"
    },
    "extreme_fg_color": {
      "color": "{Gray.1000}"
    },
    "widget_inactive_bg_fill": {
      "color": "{Gray.300}"
    },
    "widget_hovered_color": {
      "color": "{Gray.325}"
    },
    "widget_hovered_weak_bg_fill": {
      "color": "{Gray.325}"
    },
    "widget_hovered_bg_fill": {
      "color": "{Gray.325}"
    },
    "widget_active_weak_bg_fill": {
      "color": "{Gray.325}"
    },
    "widget_active_bg_fill": {
      "color": "{Gray.325}"
    },
    "widget_open_weak_bg_fill": {
      "color": "{Gray.325}"
    },
    "widget_noninteractive_weak_bg_fill": {
      "color": "{Gray.150}"
    },
    "widget_noninteractive_bg_fill": {
      "color": "{Gray.150}"
    },
    "selection_bg_fill": {
      "color": "{Blue.350}"
    },
    "selection_stroke_color": {
      "color": "#ADB8FF"
    },
    "widget_noninteractive_bg_stroke": {
      "color": "{Gray.250}"
    },
    "text_subdued": {
      "color": "{Gray.550}"
    },
    "text_default": {
      "color": "{Gray.775}"
    },
    "text_strong": {
      "color": "{Gray.1000}"
    },
    "error_fg_color": {
      "color": "#AB0116"
    },
    "warn_fg_color": {
      "color": "#FF7A0C"
    },
    "popup_shadow_color": {
      "color": "#00000080"
    },

    "density_graph_selected": {
      "color": "{Blue.800}"
    },
    "density_graph_unselected": {
      "color": "{Gray.700}"
    },

    "axis_color_x": {
      "color": "{Red.600}"
    },
    "axis_color_y": {
      "color": "{Green.600}"
    },
    "axis_color_z": {
      "color": "{Blue.600}"
    },
    "frustum_color": {
      "color": "{Gray.600}"
    },

    "list_item_selected_text": {
      "color": "{Blue.800}"
    },
    "list_item_active_text": {
      "color": "{Gray.1000}"
    },
    "list_item_noninteractive_text": {
      "color": "{Gray.550}"
    },
    "list_item_hovered_text": {
      "color": "{Gray.800}"
    },
    "list_item_default_text": {
      "color": "{Gray.700}"
    },
    "list_item_strong_text": {
      "color": "{Gray.800}"
    },
    "list_item_selected_icon": {
      "color": "{Blue.600}"
    },
    "list_item_active_icon": {
      "color": "{Gray.800}"
    },
    "list_item_hovered_icon": {
      "color": "{Gray.600}"
    },
    "list_item_default_icon": {
      "color": "{Gray.500}"
    },
    "list_item_hovered_bg": {
      "color": "{Gray.250}"
    },
    "list_item_active_bg": {
      "color": "{Gray.200}"
    },
    "list_item_collapse_default": {
      "color": "{Gray.700}"
    },
  },

  "alert_success": {
    "fill": {
       "color": "{Green.500}",
       "alpha": 50,
    },
    "stroke": {
       "color": "{Green.500}",
       "alpha": 125,
    },
    "icon": {
       "color": "{Green.500}"
    },
  },
  "alert_info": {
    "fill": {
       "color": "{Blue.500}",
       "alpha": 50,
    },
    "stroke": {
       "color": "{Blue.500}",
       "alpha": 125,
    },
    "icon": {
       "color": "{Blue.500}"
    },
  },
  "alert_warning": {
    "fill": {
       "color": "#DF7700",
       "alpha": 50,
    },
    "stroke": {
       "color": "#DF7700",
       "alpha": 125,
    },
    "icon": {
       "color": "#DF7700"
    },
  },
  "alert_error": {
    "fill": {
       "color": "{Red.500}",
       "alpha": 50,
    },
    "stroke": {
       "color": "{Red.500}",
       "alpha": 125,
    },
    "icon": {
       "color": "{Red.500}"
    },
  },

  "Global": {
    "Typography": {
      "Default": {
        "value": {
          "fontSize": "12px",
          "fontWeight": "Medium",
          "fontFamily": "Inter",
          "lineHeight": "16px",
          "letterSpacing": "-0.15px"
        },
        "description": "",
        "type": "typography"
      }
    }
  }
}
