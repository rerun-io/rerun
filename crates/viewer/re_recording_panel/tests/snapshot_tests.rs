#![cfg(feature = "testing")]

use re_entity_db::EntityDb;
use re_log_types::{StoreId, StoreKind};
use re_recording_panel::data::RecordingPanelData;
use re_test_context::TestContext;

#[test]
fn empty_context_test() {
    let test_context = TestContext::new();
    let servers = re_redap_browser::RedapServers::default();

    test_context.run_once_in_egui_central_panel(|ctx, _| {
        let data = RecordingPanelData::new(ctx, &servers, false);

        insta::assert_yaml_snapshot!(data);
    });
}

#[test]
fn fake_local_and_example_recordings_test() {
    let test_context = TestContext::new();
    let servers = re_redap_browser::RedapServers::default();

    let mut store_hub = test_context.store_hub.lock();

    // fake an example recording
    let mut example_entity_db = EntityDb::new(StoreId::new(
        StoreKind::Recording,
        "rerun_example_dna",
        "dna_rec_id",
    ));
    example_entity_db.data_source = Some(re_smart_channel::SmartChannelSource::RrdHttpStream {
        url: "https://app.rerun.io/version/nightly/examples/dna.rrd".to_owned(),
        follow: false,
    });
    store_hub.insert_entity_db(example_entity_db);

    // fake a local recording
    let mut local_entity_db = EntityDb::new(StoreId::new(
        StoreKind::Recording,
        "local_app_id",
        "local_rec_id",
    ));
    local_entity_db.data_source = Some(re_smart_channel::SmartChannelSource::Sdk);
    store_hub.insert_entity_db(local_entity_db);

    // fake a local blueprint (it should not be visible in the recording panel)
    let mut blueprint_entity_db = EntityDb::new(StoreId::new(
        StoreKind::Blueprint,
        "local_app_id",
        "local_blueprint_id",
    ));
    blueprint_entity_db.data_source = Some(re_smart_channel::SmartChannelSource::Sdk);

    // release the lock lest we deadlock
    drop(store_hub);

    test_context.run_once_in_egui_central_panel(|ctx, _| {
        let data = RecordingPanelData::new(ctx, &servers, false);

        insta::assert_yaml_snapshot!(data);
    });
}
