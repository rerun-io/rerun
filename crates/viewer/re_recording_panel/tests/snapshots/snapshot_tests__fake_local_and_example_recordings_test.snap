---
source: crates/viewer/re_recording_panel/tests/snapshot_tests.rs
expression: data
---
local_apps:
  - app_id: local_app_id
    is_active: false
    is_selected: false
    loaded_recordings:
      - entity_db:
          kind: Recording
          application_id: local_app_id
          recording_id: local_rec_id
  - app_id: test_app
    is_active: false
    is_selected: false
    loaded_recordings:
      - entity_db:
          kind: Recording
          application_id: test_app
          recording_id: "00000000000000000000000000000000"
local_tables: []
example_apps:
  - app_id: rerun_example_dna
    is_active: false
    is_selected: false
    loaded_recordings:
      - entity_db:
          kind: Recording
          application_id: rerun_example_dna
          recording_id: dna_rec_id
show_example_section: true
servers: []
loading_receivers: []
